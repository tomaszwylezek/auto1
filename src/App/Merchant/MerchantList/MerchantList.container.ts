import { connect } from 'react-redux';

import { MerchantList } from './MerchantList';

import { IMerchantListConnectedActions, IMerchantListConnectedState } from './MerchantList.types';
import { getMerchants } from 'redux/merchants/merchants.selector';

const mapStateToProps = (state): IMerchantListConnectedState => ({
  merchants: getMerchants(state)
});

const mapDispatchToProps: IMerchantListConnectedActions = {};

export const MerchantListContainer = connect(mapStateToProps, mapDispatchToProps)(MerchantList);
