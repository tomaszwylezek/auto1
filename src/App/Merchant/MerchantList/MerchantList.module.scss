.pagination :global(.page-link) {
  cursor: pointer;
}
