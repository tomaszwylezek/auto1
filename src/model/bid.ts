export interface IBid {
  id: string,
  carTitle: string,
  amount: number,
  created: string
}
