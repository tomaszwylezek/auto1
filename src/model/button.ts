export enum ButtonTheme {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  danger = 'danger',
  warning = 'warning',
  info = 'info',
  light = 'light',
  dark = 'dark',
  default = 'default',
  outlinePrimary = 'outline-primary',
  outlineSecondary = 'outline-secondary',
  outlineSuccess = 'outline-success',
  outlineDanger = 'outline-danger',
  outlineWarning = 'outline-warning',
  outlineInfo = 'outline-info',
  outlineLight = 'outline-light',
  outlineDark = 'outline-dark'
}

export enum Size {
  sm = 'sm',
  lg = 'lg'
}
