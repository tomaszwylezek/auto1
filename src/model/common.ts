import { Action } from 'redux';

export interface IAction<T> extends Action {
  payload: T;
}
