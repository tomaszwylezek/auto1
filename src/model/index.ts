export * from './store';
export * from './merchant';
export * from './bid';
export * from './formik';
