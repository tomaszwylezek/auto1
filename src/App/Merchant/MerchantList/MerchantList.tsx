import React, { FC, useState } from 'react';
import classNames from 'classnames';
import { IMerchantListProps } from 'App/Merchant/MerchantList/MerchantList.types';
import { ButtonTheme, IMerchant } from 'model';
import { Button } from 'shared/Button/Button';

import styles from './MerchantList.module.scss';
import { MerchantDetails } from 'App/Merchant/MerchantDetails/MerchantDetails';
// Models

const PER_PAGE: number = 5;

export const MerchantList: FC<IMerchantListProps> = ({ merchants, deleteMerchant }): JSX.Element => {
  const [page, setPaginationPage] = useState<number>(1);

  const setPage = (page: number) => {
    setPaginationPage(page);
    window.scrollTo(0, 0);
  };

  const currentMerchants: IMerchant[] = merchants.slice((page - 1) * PER_PAGE, page * PER_PAGE);

  return (
    <section>
      <Button theme={ButtonTheme.success} to="/add">
        Add
      </Button>
      {currentMerchants.map(
        (merchant: IMerchant): JSX.Element => {
          const name: string = `${merchant.firstname} ${merchant.lastname}`;
          const deleteSpecificMerchant: () => void = () => deleteMerchant(merchant.id);
          return (
            <MerchantDetails
              key={merchant.id}
              id={merchant.id}
              name={name}
              bids={merchant.bids}
              phone={merchant.phone}
              onDeleteClick={deleteSpecificMerchant}
            />
          );
        }
      )}
      <nav className="d-flex justify-content-center">
        <ul className={classNames('pagination', styles.pagination)}>
          <li className="page-item">
            <span className="page-link" onClick={() => setPage(Math.max(page - 1, 1))}>
              Previous
            </span>
          </li>

          {Array(Math.ceil(merchants.length / 5))
            .fill(0)
            .map((el, i) => (
              <li className="page-item" key={i}>
                <span className="page-link" onClick={() => setPage(i + 1)}>
                  {i + 1}
                </span>
              </li>
            ))}

          <li className="page-item">
            <span
              className="page-link"
              onClick={() => setPage(Math.min(page + 1, Math.ceil(merchants.length / PER_PAGE)))}
            >
              Next
            </span>
          </li>
        </ul>
      </nav>
    </section>
  );
};
