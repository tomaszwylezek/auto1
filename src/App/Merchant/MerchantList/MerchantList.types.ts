import { IMerchant } from 'model';

export interface IMerchantListOwnProps {}

export interface IMerchantListConnectedState {
  merchants: IMerchant[];
}

export interface IMerchantListConnectedActions {
  deleteMerchant: (id: string) => void;
}

export interface IMerchantListProps
  extends IMerchantListConnectedState,
    IMerchantListConnectedActions,
    IMerchantListOwnProps {}

export interface IMerchantListState {}
