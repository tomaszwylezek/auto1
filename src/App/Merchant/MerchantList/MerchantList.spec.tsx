import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { MerchantList } from './MerchantList';

// Models
import { IMerchantListProps, IMerchantListState } from './MerchantList.types';
import { IMerchant } from 'model';

describe('MerchantList component should', (): void => {
  let props: IMerchantListProps;
  let component: ShallowWrapper<IMerchantListProps, IMerchantListState>;

  beforeEach((): void => {
    props = {
      deleteMerchant: jest.fn(),
      merchants: []
    };

    window.scrollTo = jest.fn();
    component = shallow(<MerchantList {...props} />);
  });

  it('render correctly', (): void => {
    expect(component).toMatchSnapshot();
  });

  it('render correctly with merchants', (): void => {
    component.setProps({
      merchants: [
        { id: '123', firstname: 'foo', lastname: 'bar' },
        { id: '1234', firstname: 'foo', lastname: 'bar' },
        { id: '122333', firstname: 'foo', lastname: 'bar' },
        { id: '12424134', firstname: 'foo', lastname: 'bar' },
        { id: '125323', firstname: 'foo', lastname: 'bar' },
        { id: '12532334', firstname: 'foo', lastname: 'bar' }
      ] as IMerchant[]
    });
    expect(component).toMatchSnapshot();
  });

  it('render next five after page changed', (): void => {
    component.setProps({
      merchants: [
        { id: '123', firstname: 'foo', lastname: 'bar' },
        { id: '1234', firstname: 'foo', lastname: 'bar' },
        { id: '122333', firstname: 'foo', lastname: 'bar' },
        { id: '12424134', firstname: 'foo', lastname: 'bar' },
        { id: '125323', firstname: 'foo', lastname: 'bar' },
        { id: '12532334', firstname: 'foo', lastname: 'bar' }
      ] as IMerchant[]
    });

    component
      .find('.page-link')
      .at(2)
      .simulate('click');

    expect(
      component
        .find('MerchantDetails')
        .at(0)
        .prop('id')
    ).toBe('12532334');
    expect(window.scrollTo).toHaveBeenCalled();
  });
});
