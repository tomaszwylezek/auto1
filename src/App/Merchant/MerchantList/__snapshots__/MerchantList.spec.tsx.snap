// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerchantList component should render correctly 1`] = `
<section>
  <Button
    theme="success"
    to="/add"
  >
    Add
  </Button>
  <nav
    className="d-flex justify-content-center"
  >
    <ul
      className="pagination pagination"
    >
      <li
        className="page-item"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          Previous
        </span>
      </li>
      <li
        className="page-item"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          Next
        </span>
      </li>
    </ul>
  </nav>
</section>
`;

exports[`MerchantList component should render correctly with merchants 1`] = `
<section>
  <Button
    theme="success"
    to="/add"
  >
    Add
  </Button>
  <MerchantDetails
    id="123"
    key="123"
    name="foo bar"
    onDeleteClick={[Function]}
  />
  <MerchantDetails
    id="1234"
    key="1234"
    name="foo bar"
    onDeleteClick={[Function]}
  />
  <MerchantDetails
    id="122333"
    key="122333"
    name="foo bar"
    onDeleteClick={[Function]}
  />
  <MerchantDetails
    id="12424134"
    key="12424134"
    name="foo bar"
    onDeleteClick={[Function]}
  />
  <MerchantDetails
    id="125323"
    key="125323"
    name="foo bar"
    onDeleteClick={[Function]}
  />
  <nav
    className="d-flex justify-content-center"
  >
    <ul
      className="pagination pagination"
    >
      <li
        className="page-item"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          Previous
        </span>
      </li>
      <li
        className="page-item"
        key="0"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          1
        </span>
      </li>
      <li
        className="page-item"
        key="1"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          2
        </span>
      </li>
      <li
        className="page-item"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          Next
        </span>
      </li>
    </ul>
  </nav>
</section>
`;
